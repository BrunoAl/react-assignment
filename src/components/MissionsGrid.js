import React, { PureComponent } from 'react';
import { getPastLaunchesRequest } from '../requests';
import MissionCard from './MissionCard';
import '../styles/missionsGrid.css';

export default class LaunchesGrid extends PureComponent {
    componentIsmounted = false

    state = {
        launches: null,
    }

    componentDidMount() {
        this.componentIsmounted = true;
        this.getLaunches();
    }

    componentWillUnmount() {
        this.componentIsmounted = false;
        this.getLaunches();
    }

    getLaunches = () => {
        getPastLaunchesRequest().then(({ data }) => {
            if (!this.componentIsmounted) return;
            this.setState({ launches: data });
        });
    }

    render() {
        const { launches } = this.state;
        if (!launches) {
            return <h1>Loading...</h1>;
        }
        if (launches.length < 1) {
            return <h1>No Mission Found</h1>;
        }
        return (
            <section className="missions">
                {launches.map(launch => (
                    <MissionCard launch={launch} />
                ))}
            </section>
        );
    }
}
