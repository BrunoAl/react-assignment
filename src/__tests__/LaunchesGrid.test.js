import React from 'react';
import axiosMock from 'axios';

import { render, wait } from 'react-testing-library';
import missionsMock from './missions.json';
import MissionsGrid from '../components/MissionsGrid';

/**
 * Factory function to create DOM Nodes for MissionsGrid
 * @function setup
 * @param {object} props
 * @returns {DOMNode}
 */
const setupComponent = props => render(<MissionsGrid {...props} />);

describe('Missions Grid Test', () => {
    test('fetch missions and display them', async () => {
        axiosMock.get.mockResolvedValueOnce({ data: missionsMock });
        const { container, getByText } = setupComponent();

        await wait();

        expect(axiosMock.get).toHaveBeenCalled();
        expect(getByText('MissionTest').textContent).toBe('MissionTest');
        expect(container.firstChild).toMatchSnapshot();
    });

    test('Display "No Mission Found" message when there is no mission', async () => {
        axiosMock.get.mockResolvedValueOnce({ data: [] });
        const { getByText } = setupComponent();

        await wait();

        expect(axiosMock.get).toHaveBeenCalled();
        expect(getByText(/No mission found/i).textContent).toMatch(/no mission found/i);
    });

    test('Show loader when fetching data and hide it when data is fetched', async () => {
        axiosMock.get.mockResolvedValueOnce({ data: missionsMock });
        const { getByText, queryByText } = setupComponent();
        getByText(/loading.../i);
        await wait();
        expect(queryByText(/loading.../i)).toBeNull();
    });
});
