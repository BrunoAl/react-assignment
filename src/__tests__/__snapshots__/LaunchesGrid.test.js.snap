// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions Grid Test fetch missions and display them 1`] = `
<section
  class="missions"
>
  <div
    class="mission"
  >
    <img
      alt="House 1"
      class="mission__img"
      src="https://images2.imgbox.com/3c/0e/T8iJcSN3_o.png"
    />
    <h5
      class="mission__name"
    >
      MissionTest
    </h5>
    <div
      class="mission__details"
    >
      <div
        class="mission__cardDetailItem"
      >
        <img
          alt="map pin icon"
          src="mapPin.png"
        />
        <p>
          Kwajalein Atoll
        </p>
      </div>
      <div
        class="mission__cardDetailItem"
      >
        <img
          alt="rocket icon"
          src="rocket.png"
        />
        <p>
          Rocket 1
        </p>
      </div>
      <div
        class="mission__cardDetailItem"
      >
        <img
          alt="rocket icon"
          src="launchedRocket.png"
        />
        <p>
          2018
        </p>
      </div>
    </div>
    <button
      class="btn mission__btn"
      type="button"
    >
      Detalhes
    </button>
  </div>
</section>
`;
